src/main/scala/org/kframework/minikore/interfaces/MiniKoreInterface.scala

```diff
@@ -0,0 +1,303 @@
+package org.kframework.minikore.interfaces
+
+object TreeInterface {
+
+  import PatternInterface._
+
+  sealed trait AST
+
+
+  sealed trait Node extends AST with Product {
+    def args: Seq[Pattern]
+
+    def build(children: Seq[Pattern]): Pattern
+  }
+
+  object Node {
+    def unapply(arg: Node): Option[Seq[Pattern]] = Some(arg.args)
+  }
+
+
+  sealed trait Leaf[C] extends AST with Product {
+    def contents: C
+
+    def build(contents: C): Pattern
+  }
+
+  object Leaf {
+    def unapply[C](arg: Leaf[C]): Option[C] = Some(arg.contents)
+  }
+
+
+  sealed trait Leaf2[CC1, CC2] extends Leaf[Product2[CC1, CC2]] with Product2[CC1, CC2] {
+    override def contents: Product2[CC1, CC2] = (_1, _2)
+
+    def apply(_1: CC1, _2: CC2): Pattern
```

Comment:
The convention is not documented in the official style guide (which is very short) but it does appear in one very popular style guide: https://github.com/databricks/scala-style-guide#apply-method.

And it makes sense. Data structures are usually **not** constructors, at least not in our case. Let's say someone writes:
`And(a, b)(c, d)`, would it be reasonable to expect the result to be the conjunction of `a`, `b`, `c`, and `d`? Don't you find it surprising that it actually means the conjunction of just `c` and `d`? Is it good to find it surprising?

I'm not saying a non-companion object should never "apply", but it needs to make sense semantically. E.g., `Rewrite` could have the ability to apply, rewriting the target term. So one could write `Rewrite(x, Not(x))(a)` to apply the rewrite rule `x => Not(x)` to constant a, yielding `Not(a)`. But then, given we also have the weird `apply`, we can also do `Rewrite(x, Not(x))(a, a)`, which means `a => a`. So:
`Rewrite(x, Not(x))(a)` yields `Not(a)`
`Rewrite(x, Not(x))(a, a)` yields `a => a`

I find this extremely confusing.